### PR TITLE
refactor: Add tally module and refactor wasm types

### PR DESCRIPTION
**Pull request by hacheigriega**
## Motivation

This PR creates a new module `x/tally` to house the tally-related logic. Note that this module does not store any states.
This PR creates two distinct wasm types `DataRequestWasm` and `ExecutorWasm`, where the former contains pruning-related fields and the latter does not. The enum-like field `WasmType`, which used to indicate the wasm type, has been removed.

## Explanation of Changes

The term "overlay" has been renamed to "executor."
Here is the list of API changes:

- CLI endpoints
  - `sedad query wasm-storage overlay-wasm <hash>` -> `sedad query wasm-storage executor-wasm <hash>`
  - `sedad query wasm-storage list-overlay-wasms` -> `sedad query wasm-storage list-executor-wasms`
  - `sedad tx wasm-storage submit-proposal store-overlay-wasm ...` -> `sedad tx wasm-storage submit-proposal store-executor-wasm ...`
  -  `sedad tx wasm-storage submit-proposal instantiate-and-register-core-contract ...` -> `sedad tx wasm-storage submit-proposal instantiate-core-contract ...`
- REST endpoints
  - `GET /seda-chain/wasm-storage/overlay_wasms` -> `GET /seda-chain/wasm-storage/executor_wasms`
  - `GET /seda-chain/wasm-storage/overlay_wasm/{hash}` -> `GET /seda-chain/wasm-storage/executor_wasm/{hash}`
  - The response types for queries of all data request wasms and all executor (formerly overlay) wasms have changed.

## Related PRs and Issues

Closes: #309
Closes: #310

### CODE REVIEW

File: x/tally/keeper/abci.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+// to be tallied and then goes through it to filter and tally.
+func (k Keeper) ProcessTallies(ctx sdk.Context) error {
+	// Get core contract address.
+	coreContract, err := k.wasmStorageKeeper.GetCoreContractAddr(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Fetch tally-ready data requests.
+	// TODO: Deal with offset and limits.
```
Comment (Thomasvdam) [inline x/tally/keeper/abci.go:51]: Nitpick, I like to link to the issue in the TODO.

**Review by Thomasvdam — APPROVED** [review]
We should be mindful that this includes breaking changes. Is there a way we can get `Goreleaser` to add this to the release notes?

Once merged the SEDA SDK will need to update the proto definitions and generate the implementation. I think the overlay is not affected, but we should double check :)

**Review by mariocao — APPROVED** [review]

File: app/app.go [hunk 2]
```diff
@@ -617,7 +622,8 @@ func NewApp(
 		contractKeeper,
 		app.WasmKeeper,
 	)
-	wasmStorageModule := wasmstorage.NewAppModule(appCodec, app.WasmStorageKeeper)
+
+	app.TallyKeeper = tallykeeper.NewKeeper(app.WasmStorageKeeper, contractKeeper, app.WasmKeeper)
```
[the remark this answers is not in the document]
Comment (hacheigriega) [reply app/app.go:626]: Interesting question.. I was just following the SDK standard, but I see that the keeper attributes are mainly used for creating a genesis in `export.go` and for upgrade handlers in `upgrade.go`.

**Comment by hacheigriega** [comment]
Added `CHANGELOG.md`

**Pull request merged**